cli/config/config_rules.go

```diff
@@ -0,0 +1,192 @@
+		gplog.Warn("the PXF host list of the cluster `%s` will be ignored, "+
+			"because the cluster is marked as collocated", p.Name)
+	}
+	return nil
+}
+
+// A collocated PXF cluster SHOULD NOT have an endpoint, and it will be ignored
+// A warning will be generated
+func pxfClusterIgnoreEndpointWhenCollocatedWarning(p PxfCluster) error {
+	if p.Collocated && p.Endpoint != "" {
+		gplog.Warn("the PXF endpoint `%s` of the cluster `%s` will be ignored, "+
+			"because the cluster is marked as collocated", p.Endpoint, p.Name)
+	}
+	return nil
+}
+
+// An external cluster MUST have either a non-empty list of hosts or an endpoint or both
+func pxfClusterEndpointOrHostsMustExistWhenExternal(p PxfCluster) error {
+	if !p.Collocated && p.Endpoint == "" && len(p.Hosts) == 0 {
+		return fmt.Errorf("the host list and endpoint cannot be both empty for the external PXF cluster `%s`",
+			p.Name)
+	}
+	return nil
+}
+
+// A non-empty endpoint of a PXF cluster MUST have a value of either a valid IPv4 address or FQDN
+func pxfClusterValidEndpoint(p PxfCluster) error {
+	if p.Endpoint != "" && !IsValidAddress(p.Endpoint) {
+		return fmt.Errorf("the endpoint `%s` of the PXF cluster `%s` is not a valid IPv4 address "+
+			"or FQDN", p.Endpoint, p.Name)
+	}
+	return nil
+}
+
+// A PXF cluster MUST have at least one PXF service group
+func pxfClusterAtLeastOneServiceGroup(p PxfCluster) error {
+	if len(p.Groups) == 0 {
+		return fmt.Errorf("there should be at least one service group in the PXF cluster `%s`", p.Name)
+	}
+	return nil
+}
+
+// PXF service group name MUST NOT be empty
+func pxfServiceGroupNameNotEmpty(p PxfServiceGroup) error {
+	if p.Name == "" {
+		return fmt.Errorf("the name of the service group cannot be empty string")
+	}
+	return nil
+}
+
+// PXF service group MUST have more than one port
+func pxfServiceGroupAtLeastOnePort(p PxfServiceGroup) error {
+	if len(p.Ports) == 0 {
+		return fmt.Errorf("there should be at least one port in the PXF service group `%s`", p.Name)
+	}
+	return nil
+}
+
+// PXF service group MUST have ports which are integers between 1-65535
+func pxfServiceGroupValidPorts(p PxfServiceGroup) error {
+	MinPortNumber := 1
+	MaxPortNumber := 65535
+	for _, port := range p.Ports {
+		if port < MinPortNumber || port > MaxPortNumber {
+			return fmt.Errorf("invalid port number `%d` under service group `%s`, "+
+				"a port number must be between `%d` - `%d`", port, p.Name, MinPortNumber, MaxPortNumber)
+		}
+	}
+	return nil
+}
+
+// PXF service group SHOULD have ports which are integers between 1023 and 32768
+// (between privileged and ephemeral ranges)
+func pxfServiceGroupPortsWarning(p PxfServiceGroup) error {
+	MinPortNumber := 1024
+	MaxPortNumber := 32767
+	for _, port := range p.Ports {
+		if port < MinPortNumber || port > MaxPortNumber {
+			gplog.Warn("recommend using port numbers between `%d` - `%d`, "+
+				"rather than `%d` under the service group `%s`", MinPortNumber, MaxPortNumber, port, p.Name,
```

Comment:
```suggestion
			gplog.Warn("the port number provided under the service group `%s` (port: `%d`) is outside the recommended range `%d` - `%d`", p.Name, port, MinPortNumber, MaxPortNumber,
```

or something similar to this sounds a little better in my opinion